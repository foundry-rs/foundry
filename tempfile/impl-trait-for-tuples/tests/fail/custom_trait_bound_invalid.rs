trait Test {
    type Arg;

    fn test(arg: Self::Arg);
}

trait Custom {
    type NewArg;

    fn new_test(arg: Self::NewArg);
}

#[impl_trait_for_tuples::impl_for_tuples(5)]
#[tuple_types_custom_trait_bound(Custom, Clone)]
impl Test for Tuple {
    for_tuples!( type Arg = ( #( Tuple::NewArg ),* ); );
    fn test(arg: Self::Arg) {
        for_tuples!( #( Tuple::new_test(arg.Tuple); )* );
    }
}

struct Impl;

impl Custom for Impl {
    type NewArg = ();

    fn new_test(_arg: Self::NewArg) {}
}

fn test<T: Test>() {}
fn main() {
    test::<(Impl, Impl)>();
    test::<(Impl, Impl, Impl)>();
}
