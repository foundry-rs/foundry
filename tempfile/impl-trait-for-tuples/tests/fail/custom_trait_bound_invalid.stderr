error: Invalid trait bound: unexpected token
  --> tests/fail/custom_trait_bound_invalid.rs:14:40
   |
14 | #[tuple_types_custom_trait_bound(Custom, Clone)]
   |                                        ^

error[E0277]: the trait bound `(Impl, Impl): Test` is not satisfied
  --> tests/fail/custom_trait_bound_invalid.rs:32:12
   |
32 |     test::<(Impl, Impl)>();
   |            ^^^^^^^^^^^^ the trait `Test` is not implemented for `(Impl, Impl)`
   |
help: this trait has no implementations, consider adding one
  --> tests/fail/custom_trait_bound_invalid.rs:1:1
   |
1  | trait Test {
   | ^^^^^^^^^^
note: required by a bound in `test`
  --> tests/fail/custom_trait_bound_invalid.rs:30:12
   |
30 | fn test<T: Test>() {}
   |            ^^^^ required by this bound in `test`

error[E0277]: the trait bound `(Impl, Impl, Impl): Test` is not satisfied
  --> tests/fail/custom_trait_bound_invalid.rs:33:12
   |
33 |     test::<(Impl, Impl, Impl)>();
   |            ^^^^^^^^^^^^^^^^^^ the trait `Test` is not implemented for `(Impl, Impl, Impl)`
   |
help: this trait has no implementations, consider adding one
  --> tests/fail/custom_trait_bound_invalid.rs:1:1
   |
1  | trait Test {
   | ^^^^^^^^^^
note: required by a bound in `test`
  --> tests/fail/custom_trait_bound_invalid.rs:30:12
   |
30 | fn test<T: Test>() {}
   |            ^^^^ required by this bound in `test`
